walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,71 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<String> attendees = request.getAttendees();
+      Collection<String> optional = request.getOptionalAttendees();
+      List<String> newAttendees = new ArrayList<>();
+      for (String attendee: attendees) newAttendees.add(attendee);
+      for (String attendee: optional) newAttendees.add(attendee);
+      MeetingRequest newRequest = new MeetingRequest(newAttendees, request.getDuration());
+      Collection<TimeRange> withOptional = queryWithMandatoryAttendee(events, newRequest);
+      Collection<TimeRange> withoutOptional = queryWithMandatoryAttendee(events, request);
+      if (attendees.isEmpty() && !optional.isEmpty()) {
+          return withOptional;
+      }
+      if (withOptional.isEmpty()) {
+          return withoutOptional;
+      }
+      return withOptional;
+  }
```

Comment:
Yeah for this part I was always creating a new query with all the required and optional attendees marked as required because according to the requirement, if there are time slots available for both required and optional employees, then we should only return time slots available for all the required and optional employees. If there are no time slots available for both required and optional employees, then we should return the time slots available for required employees. So based on this requirement, I think the best solution is to create two queries in order to reduce code duplication.